src/htmlParser.js

```diff
@@ -42,146 +42,253 @@ const endTag = /^<\/([\-A-Za-z0-9_]+)[^>]*>/;
 
-  function append(str) {
-    stream += str;
-  }
+    const match = stream.match(startTag);
+
+    if (match) {
+      const attrs = {};
+      const booleanAttrs = {};
+      let rest = match[2];
+
+      match[2].replace(attr, function(match, name) {
+        if (!(arguments[2] || arguments[3] || arguments[4] || arguments[5])) {
+          attrs[name] = null;
+        } else if (arguments[5]) {
+          attrs[arguments[5]] = '';
+          booleanAttrs[name] = true;
+        } else {
+          attrs[name] = arguments[2] || arguments[3] || arguments[4] ||
+            fillAttr.test(name) && name || '';
+        }
+        rest = rest.replace(match, '');
+      });
 
-  function prepend(str) {
-    stream = str + stream;
+      return {
+        tagName: match[1],
+        attrs,
+        booleanAttrs,
+        rest,
+        unary: !!match[3],
+        length: match[0].length
+      };
+    }
+  },
+
+  chars: stream => {
+    const index = stream.indexOf('<');
+    return {
+      length: index >= 0 ? index : stream.length
+    };
   }
+};
+
+function fixedReadTokenFactory(parser, options) {
+  // Empty Elements - HTML 4.01
+  const EMPTY = /^(AREA|BASE|BASEFONT|BR|COL|FRAME|HR|IMG|INPUT|ISINDEX|LINK|META|PARAM|EMBED)$/i;
 
-  // Order of detection matters: detection of one can only
-  // succeed if detection of previous didn't
-  const detect = {
-    comment: /^<!--/,
-    endTag: /^<\//,
-    atomicTag: /^<\s*(script|style|noscript|iframe|textarea)[\s\/>]/i,
-    startTag: /^</,
-    chars: /^[^<]/
+  // Elements that you can| intentionally| leave open
+  // (and which close themselves)
+  const CLOSESELF = /^(COLGROUP|DD|DT|LI|OPTIONS|P|TD|TFOOT|TH|THEAD|TR)$/i;
+
+  const stack = [];
+  stack.last = function() {
+    return this[this.length - 1];
+  };
+
+  stack.lastTagNameEq = function(tagName) {
+    const last = this.last();
+    return last && last.tagName &&
+      last.tagName.toUpperCase() === tagName.toUpperCase();
   };
 
-  // Detection has already happened when a reader is called.
-  const reader = {
-    comment: () => {
-      const index = stream.indexOf('-->');
-      if (index >= 0) {
-        return {
-          content: stream.substr(4, index - 1),
-          length: index + 3
-        };
+  stack.containsTagName = function(tagName) {
+    for (let i = 0, tok; (tok = this[i]); i++) {
+      if (tok.tagName === tagName) {
+        return true;
       }
-    },
+    }
+    return false;
+  };
 
-    endTag: () => {
-      const match = stream.match(endTag);
+  function correct(tok) {
+    if (tok && tok.type === 'startTag') {
+      // unary
+      tok.unary = EMPTY.test(tok.tagName) || tok.unary;
+      tok.html5Unary = !(/\/>$/).test(tok.text);
+    }
+    return tok;
+  }
 
-      if (match) {
-        return {
-          tagName: match[1],
-          length: match[0].length
-        };
-      }
-    },
+  function peekToken() {
+    const tmp = parser.stream;
+    const tok = correct(parser._readTokenImpl());
```

Comment:
Not so great that this has to reach inside and call a private member of `parser`.